### PR TITLE
Shared Combat Infrastructure (Bonus Actions, Rest Mechanics, Events) #387

### LINKED ISSUE #387: Shared Combat Infrastructure (Bonus Actions, Rest Mechanics, Events)
## Overview
Prerequisite infrastructure needed by both Fighter (#388) and Monk (#389) level 1-3 features.

Parent issue: #386

## Why This Exists
Fighter and Monk both need:
- **Bonus action economy** - Flurry of Blows, Two-Weapon Fighting, bonus action unarmed strikes
- **Rest mechanics** - Second Wind, Ki Points, Action Surge all recover on rest
- **New event types** - Healing, resource consumption, action tracking

### Bonus Action Economy
- [ ] Track bonus action usage per turn (like how we track reactions)
- [ ] Action types on attacks (standard action vs bonus action)
- [ ] Prevent multiple bonus actions per turn

### Rest Mechanics
- [ ] Short rest event/handling
- [ ] Long rest event/handling
- [ ] Resource recovery tied to rest type (Ki = short, some features = long)

### Events Infrastructure
- [ ] `HealEvent` / `HealChain` - how healing flows from source to character
- [ ] Consider: Does healing go through a chain like damage? (resistances to healing, amplification?)
- [ ] `ResourceConsumedEvent` - Ki spent, Second Wind used, etc.
- [ ] `RestEvent` - triggers resource recovery

### Healing Flow
Currently damage flows: `AttackChain` → `DamageChain` → applied to target

For healing, consider:
- Does Second Wind just directly heal, or go through a chain?
- Are there effects that modify healing? (e.g., "healing received is halved")
- Start simple (direct heal), add chain if needed

### Resource System
We have `mechanics/resources` module. Questions:
- Should Ki, Second Wind uses, Action Surge all use this?
- Or are class-specific resources handled differently?
- Look at existing resource patterns before deciding

### String vs Ref vs Enum
When you see strings like `"bonus_action"` or `"short_rest"`, these are **placeholders, not decisions**. Feel free to:
- Use refs if it fits the pattern (`dnd5e:actions:bonus`)
- Use typed constants if simpler (`const ActionBonus ActionType = "bonus"`)
- The current codebase uses strings because we add-as-we-go, not because strings are correct

### Event Naming
Follow existing patterns in `rulebooks/dnd5e/events/`:
- `AttackTopic`, `DamageTopic` for pub/sub events
- `AttackChain`, `DamageChain` for modifier chains
- New events should follow same topology

## Acceptance Criteria
- [ ] Can track bonus action used this turn
- [ ] Can trigger short/long rest and recover appropriate resources
- [ ] Healing can be applied to characters (even if simple direct heal initially)
- [ ] Events integrate with existing event bus patterns

## Notes
- This is infrastructure - keep it minimal, don't over-engineer
- Fighter and Monk issues will drive what's actually needed
- If unsure, implement the simplest thing that works

### PR DESCRIPTION
**Pull request by KirkDiggler**
## Summary

Implements shared combat infrastructure needed by Fighter (#388) and Monk (#389) features:

- **BusEffect interface** (`core/bus_effect.go`) - Standard lifecycle for event-driven mechanics with Apply/Remove/IsApplied
- **RestEvent + RestTopic** (`rulebooks/dnd5e/events/`) - TypedTopic for rest events using existing `ResetType`
- **RecoverableResource** (`rulebooks/dnd5e/combat/`) - Wraps `mechanics/resources.Resource`, implements BusEffect, self-manages recovery by subscribing to rest events
- **Character.GetResource()** (`rulebooks/dnd5e/character/`) - First-class resource storage with serialization via `RecoverableResourceData`
- **ActionEconomy** (`rulebooks/dnd5e/combat/`) - Tracks actions/bonus/reactions with counts (not bools) to support Action Surge
- **HealChain** (`rulebooks/dnd5e/combat/`) - ChainedTopic following DamageChain pattern for modifiable healing
- **ResourceConsumedEvent** (`rulebooks/dnd5e/events/`) - TypedTopic for observability when resources are spent

## Closes

Closes #387

## Test Plan

- [x] BusEffect interface compiles and follows ConditionBehavior pattern
- [x] ActionEconomy tests pass (34 test cases covering all methods, edge cases, Action Surge scenario)
- [x] RecoverableResource tests pass (11 test cases covering Apply/Remove lifecycle, rest event filtering, resource recovery)
- [x] Character resource tests pass (11 test cases covering Add/Get, serialization round-trip)
- [x] All existing tests pass (`go test ./rulebooks/dnd5e/...`)
- [x] Linting passes (`golangci-lint run`)

## Ready For

- **Fighter (#388)**: Second Wind uses HealChain + RecoverableResource, Action Surge uses ActionEconomy
- **Monk (#389)**: Ki Points use RecoverableResource with short_rest recovery

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: rulebooks/dnd5e/character/character_test.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+		resources: nil,
+	}
+	data := char.GetResourceData()
+	s.Assert().Nil(data)
+}
+
+func (s *CharacterResourceTestSuite) TestLoadResourceDataRestoresResources() {
+	// Create data
+	data := map[coreResources.ResourceKey]RecoverableResourceData{
+		"rage": {
```
[the remark this answers is not in the document]
Comment (KirkDiggler) [reply rulebooks/dnd5e/character/character_test.go]: Also rage_charges might be better
